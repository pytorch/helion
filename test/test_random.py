from __future__ import annotations

import unittest

import torch

import helion
from helion._testing import DEVICE
from helion._testing import RefEagerTestBase
from helion._testing import TestCase
from helion._testing import code_and_output
import helion.language as hl


class TestRandom(RefEagerTestBase, TestCase):
    def test_hl_rand_1d(self):
        @helion.kernel(static_shapes=False)
        def rand_kernel_tiled_1d(x: torch.Tensor, seed: int) -> torch.Tensor:
            output = torch.zeros_like(x)
            (m,) = x.shape
            for tile_m in hl.tile(m):
                output[tile_m] = hl.rand([tile_m], seed=seed)
            return output

        x_small = torch.ones(128, device=DEVICE)
        _, output = code_and_output(rand_kernel_tiled_1d, (x_small, 42))
        _, output2 = code_and_output(rand_kernel_tiled_1d, (x_small, 1337))

        self.assertFalse(
            torch.allclose(output, output2),
            "Different seeds should produce different outputs",
        )

        code3, output3 = code_and_output(rand_kernel_tiled_1d, (x_small, 42))
        self.assertTrue(
            torch.allclose(output, output3),
            "Same seed should produce identical outputs",
        )

        # Check that all values are in [0, 1) range
        self.assertTrue(torch.all(output >= 0.0), "All values should be >= 0")
        self.assertTrue(torch.all(output < 1.0), "All values should be < 1")

        self.assertExpectedJournal(code3)

    def test_hl_rand_2d(self):
        @helion.kernel(static_shapes=False)
        def rand_kernel_tiled_2d(x: torch.Tensor, seed: int) -> torch.Tensor:
            output = torch.zeros_like(x)
            m, n = x.shape
            for tile_m, tile_n in hl.tile([m, n]):
                output[tile_m, tile_n] = hl.rand([tile_m, tile_n], seed=seed)
            return output

        x_small = torch.ones(128, 128, device=DEVICE)
        _, output = code_and_output(rand_kernel_tiled_2d, (x_small, 42))
        _, output2 = code_and_output(rand_kernel_tiled_2d, (x_small, 1337))

        self.assertFalse(
            torch.allclose(output, output2),
            "Different seeds should produce different outputs",
        )

        code3, output3 = code_and_output(rand_kernel_tiled_2d, (x_small, 42))
        self.assertTrue(
            torch.allclose(output, output3),
            "Same seed should produce identical outputs",
        )

        self.assertTrue(torch.all(output >= 0.0), "All values should be >= 0")
        self.assertTrue(torch.all(output < 1.0), "All values should be < 1")
        self.assertExpectedJournal(code3)

    def test_hl_rand_3d(self):
        @helion.kernel(static_shapes=False)
        def rand_kernel_tiled_3d(x: torch.Tensor, seed: int) -> torch.Tensor:
            output = torch.zeros_like(x)
            b, m, n = x.shape
            for tile_b, tile_m, tile_n in hl.tile([b, m, n]):
                output[tile_b, tile_m, tile_n] = hl.rand(
                    [tile_b, tile_m, tile_n], seed=seed
                )
            return output

        x_small = torch.ones(16, 32, 64, device=DEVICE)
        _, output = code_and_output(rand_kernel_tiled_3d, (x_small, 42))
        _, output2 = code_and_output(rand_kernel_tiled_3d, (x_small, 1337))

        self.assertFalse(
            torch.allclose(output, output2),
            "Different seeds should produce different outputs",
        )

        code3, output3 = code_and_output(rand_kernel_tiled_3d, (x_small, 42))
        self.assertTrue(
            torch.allclose(output, output3),
            "Same seed should produce identical outputs",
        )

        self.assertTrue(torch.all(output >= 0.0), "All values should be >= 0")
        self.assertTrue(torch.all(output < 1.0), "All values should be < 1")

        # Check distribution properties
        mean_val = output.mean().item()
        self.assertTrue(
            0.4 < mean_val < 0.6,
            f"Mean {mean_val:.3f} should be around 0.5 for uniform distribution",
        )
        self.assertExpectedJournal(code3)

    def test_hl_rand_block_size_determinism(self):
        @helion.kernel(static_shapes=False)
        def rand_kernel_2d(x: torch.Tensor, seed: int) -> torch.Tensor:
            output = torch.zeros_like(x)
            m, n = x.shape
            for tile_m, tile_n in hl.tile([m, n]):
                output[tile_m, tile_n] = hl.rand([tile_m, tile_n], seed=seed)
            return output

        x = torch.ones(128, 256, device=DEVICE)
        seed = 42

        _, output_32_32 = code_and_output(
            rand_kernel_2d, (x, seed), block_sizes=[32, 32]
        )
        _, output_64_64 = code_and_output(
            rand_kernel_2d, (x, seed), block_sizes=[64, 64]
        )
        _, output_128_128 = code_and_output(
            rand_kernel_2d, (x, seed), block_sizes=[128, 128]
        )
        _, output_16_32 = code_and_output(
            rand_kernel_2d, (x, seed), block_sizes=[16, 32]
        )

        torch.testing.assert_close(
            output_32_32,
            output_64_64,
            msg="rand should be deterministic across different block sizes (32x32 vs 64x64)",
        )
        torch.testing.assert_close(
            output_32_32,
            output_128_128,
            msg="rand should be deterministic across different block sizes (32x32 vs 128x128)",
        )
        torch.testing.assert_close(
            output_32_32,
            output_16_32,
            msg="rand should be deterministic across different block sizes (32x32 vs 16x32)",
        )

        self.assertTrue(torch.all(output_32_32 >= 0.0))
        self.assertTrue(torch.all(output_32_32 < 1.0))

    def test_hl_rand_uniqueness_distribution(self):
        @helion.kernel(static_shapes=False)
        def rand_kernel(x: torch.Tensor, seed: int) -> torch.Tensor:
            output = torch.zeros_like(x)
            m, n = x.shape
            for tile_m, tile_n in hl.tile([m, n]):
                output[tile_m, tile_n] = hl.rand([tile_m, tile_n], seed=seed)
            return output

        x = torch.ones(256, 256, device=DEVICE)
        seed = 1337

        _, output = code_and_output(rand_kernel, (x, seed))

        sorted_values = torch.sort(output.flatten()).values

        unique_values = torch.unique(sorted_values)
        total_values = output.numel()
        uniqueness_ratio = len(unique_values) / total_values

        self.assertGreater(
            uniqueness_ratio,
            0.99,
            f"Expected >99% unique values, got {uniqueness_ratio:.4f}",
        )

        n_quartile = total_values // 4
        q1_val = sorted_values[n_quartile].item()
        q2_val = sorted_values[2 * n_quartile].item()
        q3_val = sorted_values[3 * n_quartile].item()

        self.assertTrue(
            0.2 < q1_val < 0.3, f"First quartile {q1_val:.3f} should be around 0.25"
        )
        self.assertTrue(
            0.45 < q2_val < 0.55, f"Median {q2_val:.3f} should be around 0.5"
        )
        self.assertTrue(
            0.7 < q3_val < 0.8, f"Third quartile {q3_val:.3f} should be around 0.75"
        )

    def test_hl_rand_non_tiled_dimensions(self):
        @helion.kernel(static_shapes=False)
        def rand_kernel_partial_tile(x: torch.Tensor, seed: int) -> torch.Tensor:
            output = torch.zeros_like(x)
            m, n, k = x.shape
            k = hl.specialize(k)
            for tile_m, tile_n in hl.tile([m, n]):
                output[tile_m, tile_n, :] = hl.rand([tile_m, tile_n, k], seed=seed)
            return output

        x = torch.ones(64, 64, 8, device=DEVICE)
        seed = 1337

        _, output = code_and_output(rand_kernel_partial_tile, (x, seed))

        self.assertTrue(torch.all(output >= 0.0))
        self.assertTrue(torch.all(output < 1.0))

        code2, output2 = code_and_output(rand_kernel_partial_tile, (x, seed))
        torch.testing.assert_close(output, output2, msg="it should deterministic")

        self.assertExpectedJournal(code2)

    def test_hl_rand_mixed_argument_order(self):
        @helion.kernel(static_shapes=False)
        def rand_kernel_normal_order(x: torch.Tensor, seed: int) -> torch.Tensor:
            output = torch.zeros_like(x)
            m, n, k = x.shape
            for tile_m, tile_n, tile_k in hl.tile([m, n, k]):
                output[tile_m, tile_n, tile_k] = hl.rand(
                    [tile_m, tile_n, tile_k], seed=seed
                )
            return output

        @helion.kernel(static_shapes=False)
        def rand_kernel_mixed_order(x: torch.Tensor, seed: int) -> torch.Tensor:
            output = torch.zeros_like(x)
            m, n, k = x.shape
            for tile_k, tile_m, tile_n in hl.tile([k, m, n]):
                output[tile_m, tile_n, tile_k] = hl.rand(
                    [tile_m, tile_n, tile_k], seed=seed
                )
            return output

        x = torch.ones(32, 64, 16, device=DEVICE)
        seed = 1337

        code1, output1 = code_and_output(rand_kernel_normal_order, (x, seed))
        code2, output2 = code_and_output(rand_kernel_mixed_order, (x, seed))
        self.assertExpectedJournal(code1)
        self.assertExpectedJournal(code2)

        torch.testing.assert_close(
            output1,
            output2,
            msg="Mixed tile argument order should produce identical results",
        )

    def test_hl_rand_rolled_reductions(self):
        @helion.kernel(static_shapes=False)
        def rand_kernel_with_reduction(x: torch.Tensor, seed: int) -> torch.Tensor:
            m, n = x.shape
            output = torch.zeros([m], device=x.device)
            for tile_m in hl.tile(m):
                tile_values = x[tile_m, :]
                rand_values = hl.rand([tile_m], seed=seed)
                mean_val = tile_values.mean(-1)
                output[tile_m] = rand_values * mean_val
            return output

        x = torch.ones(64, 128, device=DEVICE)
        seed = 42

        code1, output_persistent = code_and_output(
            rand_kernel_with_reduction,
            (x, seed),
            block_sizes=[32],
            reduction_loops=[None],
        )
        code2, output_rolled = code_and_output(
            rand_kernel_with_reduction,
            (x, seed),
            block_sizes=[32],
            reduction_loops=[64],
        )
        self.assertExpectedJournal(code1)
        self.assertExpectedJournal(code2)

        torch.testing.assert_close(
            output_persistent,
            output_rolled,
            msg="Persistent and rolled reductions should produce identical results",
        )


if __name__ == "__main__":
    unittest.main()
