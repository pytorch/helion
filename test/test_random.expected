This file is automatically generated by assertExpectedJournal calls in test_random.py.
Update expected outputs by running tests with the EXPECTTEST_ACCEPT=1 environment variable set.

--- assertExpectedJournal(TestRandom.test_hl_rand_1d)
from __future__ import annotations

import torch
import triton
import triton.language as tl
from helion.runtime import default_launcher as _default_launcher

@triton.jit
def _helion_rand_kernel_tiled_1d(output, output_stride_0, m, seed, _BLOCK_SIZE_0: tl.constexpr):
    # src[test_random.py:N]: for tile_m in hl.tile(m):
    pid_0 = tl.program_id(0)
    offset_0 = pid_0 * _BLOCK_SIZE_0
    indices_0 = (offset_0 + tl.arange(0, _BLOCK_SIZE_0)).to(tl.int32)
    mask_0 = indices_0 < m
    # src[test_random.py:N]: output[tile_m] = hl.rand([tile_m], seed=seed)
    rand = tl.rand(seed, indices_0)
    tl.store(output + indices_0 * output_stride_0, rand, mask_0)

def rand_kernel_tiled_1d(x: torch.Tensor, seed: int, *, _launcher=_default_launcher):
    # src[test_random.py:N]: output = torch.zeros_like(x)
    output = torch.zeros_like(x)
    # src[test_random.py:N]: (m,) = x.shape
    m, = x.shape
    # src[test_random.py:N]: for tile_m in hl.tile(m):
    _BLOCK_SIZE_0 = 32
    # src[test_random.py:N]: for tile_m in hl.tile(m):
    # src[test_random.py:N]:     output[tile_m] = hl.rand([tile_m], seed=seed)
    _launcher(_helion_rand_kernel_tiled_1d, (triton.cdiv(m, _BLOCK_SIZE_0),), output, output.stride(0), m, seed, _BLOCK_SIZE_0, num_warps=4, num_stages=1)
    # src[test_random.py:N]: return output
    return output

--- assertExpectedJournal(TestRandom.test_hl_rand_2d)
from __future__ import annotations

import torch
import triton
import triton.language as tl
from helion.runtime import default_launcher as _default_launcher

@triton.jit
def _helion_rand_kernel_tiled_2d(output, output_stride_0, output_stride_1, m, n, seed, _BLOCK_SIZE_0: tl.constexpr, _BLOCK_SIZE_1: tl.constexpr):
    # src[test_random.py:N]: for tile_m, tile_n in hl.tile([m, n]):
    num_blocks_0 = tl.cdiv(m, _BLOCK_SIZE_0)
    pid_0 = tl.program_id(0) % num_blocks_0
    pid_1 = tl.program_id(0) // num_blocks_0
    offset_0 = pid_0 * _BLOCK_SIZE_0
    indices_0 = (offset_0 + tl.arange(0, _BLOCK_SIZE_0)).to(tl.int32)
    mask_0 = indices_0 < m
    offset_1 = pid_1 * _BLOCK_SIZE_1
    indices_1 = (offset_1 + tl.arange(0, _BLOCK_SIZE_1)).to(tl.int32)
    mask_1 = indices_1 < n
    # src[test_random.py:N]: output[tile_m, tile_n] = hl.rand([tile_m, tile_n], seed=seed)
    rand = tl.rand(seed, indices_0[:, None] * n + indices_1[None, :])
    tl.store(output + (indices_0[:, None] * output_stride_0 + indices_1[None, :] * output_stride_1), rand, mask_0[:, None] & mask_1[None, :])

def rand_kernel_tiled_2d(x: torch.Tensor, seed: int, *, _launcher=_default_launcher):
    # src[test_random.py:N]: output = torch.zeros_like(x)
    output = torch.zeros_like(x)
    # src[test_random.py:N]: m, n = x.shape
    m, n = x.shape
    # src[test_random.py:N]: for tile_m, tile_n in hl.tile([m, n]):
    _BLOCK_SIZE_0 = 32
    _BLOCK_SIZE_1 = 32
    # src[test_random.py:N]: for tile_m, tile_n in hl.tile([m, n]):
    # src[test_random.py:N]:     output[tile_m, tile_n] = hl.rand([tile_m, tile_n], seed=seed)
    _launcher(_helion_rand_kernel_tiled_2d, (triton.cdiv(m, _BLOCK_SIZE_0) * triton.cdiv(n, _BLOCK_SIZE_1),), output, output.stride(0), output.stride(1), m, n, seed, _BLOCK_SIZE_0, _BLOCK_SIZE_1, num_warps=4, num_stages=1)
    # src[test_random.py:N]: return output
    return output

--- assertExpectedJournal(TestRandom.test_hl_rand_3d)
from __future__ import annotations

import torch
import triton
import triton.language as tl
from helion.runtime import default_launcher as _default_launcher

@triton.jit
def _helion_rand_kernel_tiled_3d(output, output_stride_0, output_stride_1, output_stride_2, b, m, n, seed, _BLOCK_SIZE_0: tl.constexpr, _BLOCK_SIZE_1: tl.constexpr, _BLOCK_SIZE_2: tl.constexpr):
    # src[test_random.py:N]: for tile_b, tile_m, tile_n in hl.tile([b, m, n]):
    num_blocks_0 = tl.cdiv(b, _BLOCK_SIZE_0)
    num_blocks_1 = tl.cdiv(m, _BLOCK_SIZE_1)
    pid_0 = tl.program_id(0) % num_blocks_0
    pid_1 = tl.program_id(0) // num_blocks_0 % num_blocks_1
    pid_2 = tl.program_id(0) // (num_blocks_0 * num_blocks_1)
    offset_0 = pid_0 * _BLOCK_SIZE_0
    indices_0 = (offset_0 + tl.arange(0, _BLOCK_SIZE_0)).to(tl.int32)
    mask_0 = indices_0 < b
    offset_1 = pid_1 * _BLOCK_SIZE_1
    indices_1 = (offset_1 + tl.arange(0, _BLOCK_SIZE_1)).to(tl.int32)
    mask_1 = indices_1 < m
    offset_2 = pid_2 * _BLOCK_SIZE_2
    indices_2 = (offset_2 + tl.arange(0, _BLOCK_SIZE_2)).to(tl.int32)
    mask_2 = indices_2 < n
    # src[test_random.py:N]: output[tile_b, tile_m, tile_n] = hl.rand(
    # src[test_random.py:N]:     [tile_b, tile_m, tile_n], seed=seed
    # src[test_random.py:N]: )
    rand = tl.rand(seed, indices_0[:, None, None] * m * n + indices_1[None, :, None] * n + indices_2[None, None, :])
    tl.store(output + (indices_0[:, None, None] * output_stride_0 + indices_1[None, :, None] * output_stride_1 + indices_2[None, None, :] * output_stride_2), rand, mask_0[:, None, None] & mask_1[None, :, None] & mask_2[None, None, :])

def rand_kernel_tiled_3d(x: torch.Tensor, seed: int, *, _launcher=_default_launcher):
    # src[test_random.py:N]: output = torch.zeros_like(x)
    output = torch.zeros_like(x)
    # src[test_random.py:N]: b, m, n = x.shape
    b, m, n = x.shape
    # src[test_random.py:N]: for tile_b, tile_m, tile_n in hl.tile([b, m, n]):
    _BLOCK_SIZE_0 = 16
    _BLOCK_SIZE_1 = 16
    _BLOCK_SIZE_2 = 16
    # src[test_random.py:N]: for tile_b, tile_m, tile_n in hl.tile([b, m, n]):
    # src[test_random.py:N]:     output[tile_b, tile_m, tile_n] = hl.rand(
    # src[test_random.py:N]:         [tile_b, tile_m, tile_n], seed=seed
    # src[test_random.py:N-N]: ...
    _launcher(_helion_rand_kernel_tiled_3d, (triton.cdiv(b, _BLOCK_SIZE_0) * triton.cdiv(m, _BLOCK_SIZE_1) * triton.cdiv(n, _BLOCK_SIZE_2),), output, output.stride(0), output.stride(1), output.stride(2), b, m, n, seed, _BLOCK_SIZE_0, _BLOCK_SIZE_1, _BLOCK_SIZE_2, num_warps=4, num_stages=1)
    # src[test_random.py:N]: return output
    return output

--- assertExpectedJournal(TestRandom.test_hl_rand_mixed_argument_order)
from __future__ import annotations

import torch
import triton
import triton.language as tl
from helion.runtime import default_launcher as _default_launcher

@triton.jit
def _helion_rand_kernel_normal_order(output, output_stride_0, output_stride_1, output_stride_2, m, n, k, seed, _BLOCK_SIZE_0: tl.constexpr, _BLOCK_SIZE_1: tl.constexpr, _BLOCK_SIZE_2: tl.constexpr):
    # src[test_random.py:N]: for tile_m, tile_n, tile_k in hl.tile([m, n, k]):
    num_blocks_0 = tl.cdiv(m, _BLOCK_SIZE_0)
    num_blocks_1 = tl.cdiv(n, _BLOCK_SIZE_1)
    pid_0 = tl.program_id(0) % num_blocks_0
    pid_1 = tl.program_id(0) // num_blocks_0 % num_blocks_1
    pid_2 = tl.program_id(0) // (num_blocks_0 * num_blocks_1)
    offset_0 = pid_0 * _BLOCK_SIZE_0
    indices_0 = (offset_0 + tl.arange(0, _BLOCK_SIZE_0)).to(tl.int32)
    mask_0 = indices_0 < m
    offset_1 = pid_1 * _BLOCK_SIZE_1
    indices_1 = (offset_1 + tl.arange(0, _BLOCK_SIZE_1)).to(tl.int32)
    mask_1 = indices_1 < n
    offset_2 = pid_2 * _BLOCK_SIZE_2
    indices_2 = (offset_2 + tl.arange(0, _BLOCK_SIZE_2)).to(tl.int32)
    mask_2 = indices_2 < k
    # src[test_random.py:N]: output[tile_m, tile_n, tile_k] = hl.rand(
    # src[test_random.py:N]:     [tile_m, tile_n, tile_k], seed=seed
    # src[test_random.py:N]: )
    rand = tl.rand(seed, indices_0[:, None, None] * n * k + indices_1[None, :, None] * k + indices_2[None, None, :])
    tl.store(output + (indices_0[:, None, None] * output_stride_0 + indices_1[None, :, None] * output_stride_1 + indices_2[None, None, :] * output_stride_2), rand, mask_0[:, None, None] & mask_1[None, :, None] & mask_2[None, None, :])

def rand_kernel_normal_order(x: torch.Tensor, seed: int, *, _launcher=_default_launcher):
    # src[test_random.py:N]: output = torch.zeros_like(x)
    output = torch.zeros_like(x)
    # src[test_random.py:N]: m, n, k = x.shape
    m, n, k = x.shape
    # src[test_random.py:N]: for tile_m, tile_n, tile_k in hl.tile([m, n, k]):
    _BLOCK_SIZE_0 = 16
    _BLOCK_SIZE_1 = 16
    _BLOCK_SIZE_2 = 16
    # src[test_random.py:N]: for tile_m, tile_n, tile_k in hl.tile([m, n, k]):
    # src[test_random.py:N]:     output[tile_m, tile_n, tile_k] = hl.rand(
    # src[test_random.py:N]:         [tile_m, tile_n, tile_k], seed=seed
    # src[test_random.py:N-N]: ...
    _launcher(_helion_rand_kernel_normal_order, (triton.cdiv(m, _BLOCK_SIZE_0) * triton.cdiv(n, _BLOCK_SIZE_1) * triton.cdiv(k, _BLOCK_SIZE_2),), output, output.stride(0), output.stride(1), output.stride(2), m, n, k, seed, _BLOCK_SIZE_0, _BLOCK_SIZE_1, _BLOCK_SIZE_2, num_warps=4, num_stages=1)
    # src[test_random.py:N]: return output
    return output

--- assertExpectedJournal(TestRandom.test_hl_rand_mixed_argument_order)
from __future__ import annotations

import torch
import triton
import triton.language as tl
from helion.runtime import default_launcher as _default_launcher

@triton.jit
def _helion_rand_kernel_mixed_order(output, output_stride_0, output_stride_1, output_stride_2, k, m, n, seed, _BLOCK_SIZE_0: tl.constexpr, _BLOCK_SIZE_1: tl.constexpr, _BLOCK_SIZE_2: tl.constexpr):
    # src[test_random.py:N]: for tile_k, tile_m, tile_n in hl.tile([k, m, n]):
    num_blocks_0 = tl.cdiv(k, _BLOCK_SIZE_0)
    num_blocks_1 = tl.cdiv(m, _BLOCK_SIZE_1)
    pid_0 = tl.program_id(0) % num_blocks_0
    pid_1 = tl.program_id(0) // num_blocks_0 % num_blocks_1
    pid_2 = tl.program_id(0) // (num_blocks_0 * num_blocks_1)
    offset_0 = pid_0 * _BLOCK_SIZE_0
    indices_0 = (offset_0 + tl.arange(0, _BLOCK_SIZE_0)).to(tl.int32)
    mask_0 = indices_0 < k
    offset_1 = pid_1 * _BLOCK_SIZE_1
    indices_1 = (offset_1 + tl.arange(0, _BLOCK_SIZE_1)).to(tl.int32)
    mask_1 = indices_1 < m
    offset_2 = pid_2 * _BLOCK_SIZE_2
    indices_2 = (offset_2 + tl.arange(0, _BLOCK_SIZE_2)).to(tl.int32)
    mask_2 = indices_2 < n
    # src[test_random.py:N]: output[tile_m, tile_n, tile_k] = hl.rand(
    # src[test_random.py:N]:     [tile_m, tile_n, tile_k], seed=seed
    # src[test_random.py:N]: )
    rand = tl.rand(seed, indices_1[:, None, None] * n * k + indices_2[None, :, None] * k + indices_0[None, None, :])
    tl.store(output + (indices_1[:, None, None] * output_stride_0 + indices_2[None, :, None] * output_stride_1 + indices_0[None, None, :] * output_stride_2), rand, mask_1[:, None, None] & mask_2[None, :, None] & mask_0[None, None, :])

def rand_kernel_mixed_order(x: torch.Tensor, seed: int, *, _launcher=_default_launcher):
    # src[test_random.py:N]: output = torch.zeros_like(x)
    output = torch.zeros_like(x)
    # src[test_random.py:N]: m, n, k = x.shape
    m, n, k = x.shape
    # src[test_random.py:N]: for tile_k, tile_m, tile_n in hl.tile([k, m, n]):
    _BLOCK_SIZE_0 = 16
    _BLOCK_SIZE_1 = 16
    _BLOCK_SIZE_2 = 16
    # src[test_random.py:N]: for tile_k, tile_m, tile_n in hl.tile([k, m, n]):
    # src[test_random.py:N]:     output[tile_m, tile_n, tile_k] = hl.rand(
    # src[test_random.py:N]:         [tile_m, tile_n, tile_k], seed=seed
    # src[test_random.py:N-N]: ...
    _launcher(_helion_rand_kernel_mixed_order, (triton.cdiv(k, _BLOCK_SIZE_0) * triton.cdiv(m, _BLOCK_SIZE_1) * triton.cdiv(n, _BLOCK_SIZE_2),), output, output.stride(0), output.stride(1), output.stride(2), k, m, n, seed, _BLOCK_SIZE_0, _BLOCK_SIZE_1, _BLOCK_SIZE_2, num_warps=4, num_stages=1)
    # src[test_random.py:N]: return output
    return output

--- assertExpectedJournal(TestRandom.test_hl_rand_non_tiled_dimensions)
from __future__ import annotations

import torch
import triton
import triton.language as tl
from helion.runtime import default_launcher as _default_launcher

@triton.jit
def _helion_rand_kernel_partial_tile(output, output_stride_0, output_stride_1, output_stride_2, m, n, seed, _BLOCK_SIZE_0: tl.constexpr, _BLOCK_SIZE_1: tl.constexpr, _RDIM_SIZE_2: tl.constexpr):
    # src[test_random.py:N]: for tile_m, tile_n in hl.tile([m, n]):
    num_blocks_0 = tl.cdiv(m, _BLOCK_SIZE_0)
    pid_0 = tl.program_id(0) % num_blocks_0
    pid_1 = tl.program_id(0) // num_blocks_0
    offset_0 = pid_0 * _BLOCK_SIZE_0
    indices_0 = (offset_0 + tl.arange(0, _BLOCK_SIZE_0)).to(tl.int32)
    mask_0 = indices_0 < m
    offset_1 = pid_1 * _BLOCK_SIZE_1
    indices_1 = (offset_1 + tl.arange(0, _BLOCK_SIZE_1)).to(tl.int32)
    mask_1 = indices_1 < n
    indices_2 = tl.arange(0, _RDIM_SIZE_2).to(tl.int32)
    # src[test_random.py:N]: output[tile_m, tile_n, :] = hl.rand([tile_m, tile_n, k], seed=seed)
    rand = tl.rand(seed, indices_0[:, None, None] * n * _RDIM_SIZE_2 + indices_1[None, :, None] * _RDIM_SIZE_2 + indices_2[None, None, :])
    tl.store(output + (indices_0[:, None, None] * output_stride_0 + indices_1[None, :, None] * output_stride_1 + indices_2[None, None, :] * output_stride_2), rand, mask_0[:, None, None] & mask_1[None, :, None])

def rand_kernel_partial_tile(x: torch.Tensor, seed: int, *, _launcher=_default_launcher):
    # src[test_random.py:N]: output = torch.zeros_like(x)
    output = torch.zeros_like(x)
    # src[test_random.py:N]: m, n, k = x.shape
    m, n, k = x.shape
    # src[test_random.py:N]: for tile_m, tile_n in hl.tile([m, n]):
    _BLOCK_SIZE_0 = 32
    _BLOCK_SIZE_1 = 32
    _RDIM_SIZE_2 = 8
    # src[test_random.py:N]: for tile_m, tile_n in hl.tile([m, n]):
    # src[test_random.py:N]:     output[tile_m, tile_n, :] = hl.rand([tile_m, tile_n, k], seed=seed)
    _launcher(_helion_rand_kernel_partial_tile, (triton.cdiv(m, _BLOCK_SIZE_0) * triton.cdiv(n, _BLOCK_SIZE_1),), output, output.stride(0), output.stride(1), output.stride(2), m, n, seed, _BLOCK_SIZE_0, _BLOCK_SIZE_1, _RDIM_SIZE_2, num_warps=4, num_stages=1)
    # src[test_random.py:N]: return output
    return output

--- assertExpectedJournal(TestRandom.test_hl_rand_rolled_reductions)
from __future__ import annotations

import torch
import triton
import triton.language as tl
from helion.runtime import default_launcher as _default_launcher

@triton.jit
def _helion_rand_kernel_with_reduction(x, output, output_stride_0, x_stride_0, x_stride_1, m, n, seed, _BLOCK_SIZE_0: tl.constexpr, _RDIM_SIZE_1: tl.constexpr):
    # src[test_random.py:N]: for tile_m in hl.tile(m):
    pid_0 = tl.program_id(0)
    offset_0 = pid_0 * _BLOCK_SIZE_0
    indices_0 = (offset_0 + tl.arange(0, _BLOCK_SIZE_0)).to(tl.int32)
    mask_0 = indices_0 < m
    indices_1 = tl.arange(0, _RDIM_SIZE_1).to(tl.int32)
    mask_1 = indices_1 < n
    # src[test_random.py:N]: tile_values = x[tile_m, :]
    tile_values = tl.load(x + (indices_0[:, None] * x_stride_0 + indices_1[None, :] * x_stride_1), mask_0[:, None] & mask_1[None, :], other=0)
    # src[test_random.py:N]: rand_values = hl.rand([tile_m], seed=seed)
    rand_values = tl.rand(seed, indices_0)
    # src[test_random.py:N]: mean_val = tile_values.mean(-1)
    mean_val_extra = tl.cast(tl.sum(tile_values, 1), tl.float32)
    v_0 = mean_val_extra / n.to(tl.float32)
    # src[test_random.py:N]: output[tile_m] = rand_values * mean_val
    v_1 = rand_values * v_0
    tl.store(output + indices_0 * output_stride_0, v_1, mask_0)

def rand_kernel_with_reduction(x: torch.Tensor, seed: int, *, _launcher=_default_launcher):
    # src[test_random.py:N]: m, n = x.shape
    m, n = x.shape
    # src[test_random.py:N]: output = torch.zeros([m], device=x.device)
    output = torch.zeros([m], device=x.device)
    # src[test_random.py:N]: for tile_m in hl.tile(m):
    _BLOCK_SIZE_0 = 32
    _RDIM_SIZE_1 = triton.next_power_of_2(n)
    # src[test_random.py:N]: for tile_m in hl.tile(m):
    # src[test_random.py:N]:     tile_values = x[tile_m, :]
    # src[test_random.py:N]:     rand_values = hl.rand([tile_m], seed=seed)
    # src[test_random.py:N-N]: ...
    _launcher(_helion_rand_kernel_with_reduction, (triton.cdiv(m, _BLOCK_SIZE_0),), x, output, output.stride(0), x.stride(0), x.stride(1), m, n, seed, _BLOCK_SIZE_0, _RDIM_SIZE_1, num_warps=4, num_stages=1)
    # src[test_random.py:N]: return output
    return output

--- assertExpectedJournal(TestRandom.test_hl_rand_rolled_reductions)
from __future__ import annotations

import torch
import triton
import triton.language as tl
from helion.runtime import default_launcher as _default_launcher

@triton.jit
def _helion_rand_kernel_with_reduction(x, output, output_stride_0, x_stride_0, x_stride_1, m, seed, n, _BLOCK_SIZE_0: tl.constexpr, _REDUCTION_BLOCK_1: tl.constexpr):
    # src[test_random.py:N]: for tile_m in hl.tile(m):
    pid_0 = tl.program_id(0)
    offset_0 = pid_0 * _BLOCK_SIZE_0
    indices_0 = (offset_0 + tl.arange(0, _BLOCK_SIZE_0)).to(tl.int32)
    mask_0 = indices_0 < m
    # src[test_random.py:N]: rand_values = hl.rand([tile_m], seed=seed)
    rand_values = tl.rand(seed, indices_0)
    # src[test_random.py:N]: mean_val = tile_values.mean(-1)
    mean_val_extra_acc = tl.full([_BLOCK_SIZE_0, _REDUCTION_BLOCK_1], 0, tl.float32)
    # src[test_random.py:N]: tile_values = x[tile_m, :]
    for roffset_1 in tl.range(0, n, _REDUCTION_BLOCK_1):
        rindex_1 = roffset_1 + tl.arange(0, _REDUCTION_BLOCK_1).to(tl.int32)
        mask_1 = rindex_1 < n
        tile_values = tl.load(x + (indices_0[:, None] * x_stride_0 + rindex_1[None, :] * x_stride_1), mask_0[:, None] & mask_1[None, :], other=0)
        # src[test_random.py:N]: mean_val = tile_values.mean(-1)
        v_0 = mean_val_extra_acc + tile_values
        mean_val_extra_acc = v_0
    mean_val_extra = tl.cast(tl.sum(mean_val_extra_acc, 1), tl.float32)
    v_1 = mean_val_extra / n.to(tl.float32)
    # src[test_random.py:N]: output[tile_m] = rand_values * mean_val
    v_2 = rand_values * v_1
    tl.store(output + indices_0 * output_stride_0, v_2, mask_0)

def rand_kernel_with_reduction(x: torch.Tensor, seed: int, *, _launcher=_default_launcher):
    # src[test_random.py:N]: m, n = x.shape
    m, n = x.shape
    # src[test_random.py:N]: output = torch.zeros([m], device=x.device)
    output = torch.zeros([m], device=x.device)
    # src[test_random.py:N]: for tile_m in hl.tile(m):
    _BLOCK_SIZE_0 = 32
    # src[test_random.py:N]: tile_values = x[tile_m, :]
    _REDUCTION_BLOCK_1 = 64
    # src[test_random.py:N]: for tile_m in hl.tile(m):
    # src[test_random.py:N]:     tile_values = x[tile_m, :]
    # src[test_random.py:N]:     rand_values = hl.rand([tile_m], seed=seed)
    # src[test_random.py:N-N]: ...
    _launcher(_helion_rand_kernel_with_reduction, (triton.cdiv(m, _BLOCK_SIZE_0),), x, output, output.stride(0), x.stride(0), x.stride(1), m, seed, n, _BLOCK_SIZE_0, _REDUCTION_BLOCK_1, num_warps=4, num_stages=1)
    # src[test_random.py:N]: return output
    return output
